lib/api/links_builder.rb

```diff
@@ -0,0 +1,69 @@
+module Api
+  class LinksBuilder
+    def initialize(params, href, counts)
+      @offset = params["offset"].to_i if params["offset"]
+      @limit = params["limit"].to_i if params["limit"]
+      @href = href
+      return unless counts
+      @counts = counts
+    end
+
+    def links
+      {
+        :self     => self_href,
+        :next     => next_href,
+        :previous => previous_href,
+        :first    => first_href,
+        :last     => last_href
+      }.compact
+    end
+
+    def pages
+      (paging_count / limit.to_f).ceil
+    end
+
+    def links?
+      offset && limit
+    end
+
+    private
+
+    attr_reader :offset, :limit, :href, :counts
+
+    def self_href
+      @self_href ||= format_href(offset)
+    end
+
+    def format_href(new_offset)
+      href.sub "offset=#{offset}", "offset=#{new_offset}"
+    end
+
+    def paging_count
+      @paging_count ||= counts.subquery_count || counts.count
+    end
+
+    def next_href
+      next_offset = offset + limit
+      return if next_offset >= paging_count
+      format_href(next_offset)
+    end
+
+    def previous_href
+      return if offset.zero?
+      prev_offset = offset - limit
+      return unless prev_offset > (0 - limit)
+      format_href(prev_offset)
+    end
+
+    def first_href
+      return if offset.zero?
+      format_href(0)
+    end
+
+    def last_href
+      return if (offset + limit) >= paging_count
+      last_offset = paging_count - (paging_count % limit)
+      format_href(last_offset)
```

Comment:
while I can see us skipping previous and next (i.e you're in the first page already or you're in the last so no need to paginate further), I think we should always return the self, first and last href's.